torchvision/models/_api.py

```diff
@@ -140,3 +143,53 @@ def _get_enum_from_fn(fn: Callable) -> WeightsEnum:
         )
 
     return cast(WeightsEnum, weights_enum)
+
+
+M = TypeVar("M", bound=Type[nn.Module])
+
+BUILTIN_MODELS = {}
+
+
+def register_model(name: str, overwrite: bool = False) -> Callable[[Callable[..., M]], Callable[..., M]]:
+    def wrapper(fn: Callable[..., M]) -> Callable[..., M]:
+        if name in BUILTIN_MODELS and not overwrite:
+            raise ValueError(f"An entry is already registered under the name '{name}'.")
+        BUILTIN_MODELS[name] = fn
+        return fn
+
+    return wrapper
+
+
+def list_models(module: Optional[ModuleType] = None) -> List[str]:
+    """
+    Returns a list with the names of registered models.
+
+    Args:
+        module (ModuleType, optional): The module from which we want to extract the available models.
+
+    Returns:
+        models (list): A list with the names of available models.
+    """
+    models = [
+        k for k, v in BUILTIN_MODELS.items() if module is None or v.__module__.rsplit(".", 1)[0] == module.__name__
```

Comment:
Is this the same?

```suggestion
        k for k, v in BUILTIN_MODELS.items() if module is None or v.__module__ is module
```

Reply:
No it's not. `__module__` is string while `module` is an actual module. We do the operations on string level to avoid extra method calls on `inspect.getmodule(v)` or `sys.modules[v.__module__]`. If you have lightweight alternatives happy to adopt.

Reply:
My only concern is that this seems to be based on the assumption that the model builders are always defined in `THE_MODULE/some_file.py` (e.g. `models/resnet.py`) which may or may not always be the case in the future. But maybe this isn't too important

Reply:
True, doing string manipulation is not great. Maybe doing `sys.modules[v.__module__] is module` is lightweight enough. Let me give it a try.

Reply:
@NicolasHug here is what we can do:
```
k for k, v in BUILTIN_MODELS.items() if module is None or sys.modules[v.__module__].__package__ == module.__package__
```

Note that the comparison again drops on string level. Also I don't think this is actually avoid the problem that you highlighted. Perhaps we can get around it if we start calling `inspect.getmodule()` but I think that's going to be an overkill. :(

Reply:
Ah. I guess the question is: if we write a new model and it doesn't get registered properly, will we 100% catch it?

If the answer is yes we can sleep easy.

Regardless, I'd say let's just keep the implementation the way it is. It's still a rather unlikely that this won't work.